xonsh/ptk_shell/formatter.py

```diff
@@ -0,0 +1,62 @@
+import builtins
+import functools
+from concurrent.futures import ThreadPoolExecutor
+
+from xonsh.prompt.base import PromptFormatter, DEFAULT_PROMPT
+
+
+class PTKPromptFormatter(PromptFormatter):
+    def __init__(self):
+        super().__init__()
+
+        # the attribute, .cache is cleared between calls. This caches results from callback alone by field name.
+        self.thread_results = {}
+
+    def __call__(self, template=DEFAULT_PROMPT, fields=None, threaded=False):
+        """Formats a xonsh prompt template string."""
+
+        # set these attributes per call. one can enable/disable async-prompt inside a session.
+        self.template = template  # current template
+        self.threaded = threaded
+        self.futures = {}
+
+        # init pool only on async prompts
+        if threaded and (not hasattr(self, "thread_pool")):
+            self.thread_pool = ThreadPoolExecutor(
+                max_workers=builtins.__xonsh__.env["ASYNC_PROMPT_THREAD_WORKERS"]
+            )
+
+        prompt = super().__call__(template, fields)
+
+        if threaded:
+            return prompt, self.futures
+        return prompt
+
+    def _format_prompt(self, template=DEFAULT_PROMPT):
+        tokens = self._get_tokens(template)
+        if self.threaded:
+            return tokens
+        return "".join(tokens)
+
+    def _no_cache_field_value(self, field, field_value, idx: int = None):
+        func = functools.partial(super()._no_cache_field_value, field, field_value, idx)
+
+        if self.threaded and callable(field_value):
+            future = self.thread_pool.submit(self._run_func, func, field)
+
+            res = "{%s}" % field
+            if idx is None:
+                self.futures = res
+            else:
+                self.futures[future] = idx
+
+            # todo: the old result should be shown with faded colour.
+            return self.thread_results[field] if field in self.thread_results else res
```

Comment:
@BYK do you know how to style the token faded ? even when it  is surrounded by other colors ?

Reply:
It would have to be the inner-most color code, which might be hard.

Reply:
perhaps, can we wrap these fields into a class? like this

```
Section:
  content = "text"
  style = "color"
  faded_style = "color"
```

later we can select the styles based on the place. What do you think?

Reply:
I have removed any color and made it to be GRAY color.